Upgrade react-intl

renchap opened the pull request:
See the various commits for details of what is does.

TODO:
- [x] Ensure that the Crowdin process still works
- [x] Replace `manage:translations`
- [x] Ensure polyfills are loaded correctly (https://formatjs.io/docs/polyfills)
- [x] See if we can replace `<RelativeTimestamp>` with something simpler based on `react-intl` components or helpers, but I am not sure what all the props do. **status**: should be in a future PR, not really needed right now
- [x] Correctly handle `oc` and `sa` locales (see below)
- [x] Ensure no errors occur when we try to load polyfills for a non-existing locale (`oc`). **status**: `shoudPolyfillPluralRules` has a built-in list of supported locales, and fallbacks to `en`, so all here.

## Custom plural rules issue

Currently we are using `sa` and `oc` as custom plural rules, but they are not supported by the Javascript Intl API, resulting is this message on page load:
`formatjs/intl Error MISSING_DATA] Missing locale data for locale: "oc" in Intl.NumberFormat. Using default locale: "en" as fallback. See https://formatjs.io/docs/react-intl#runtime-requirements for more details`

I am not sure how to properly handle this.

The browser is supposed to pick the best fit for the provided locale, so I guess we could just make FormatJS not print this message (using `onError`) for those 2 locales?


### ✅ Solution

`oc` is not supported in Chrome & Firefox, `sa` is not supported in Chrome, both are supported in Safari. According to the JS spec, the browsers should pick the best available locale here, so we can simply silence the message.

## Fixes

- #22760

renchap committed:
Upgrade to latest react-intl

- It now uses the browser's built-in plural rules, no longer need to load them
- Translations are loaded using dynamic import, this is more future-proof

renchap committed:
Upgrade various Intl packages

renchap committed:
Remove usage of `FormattedHTMLMessage`

Its now supported by `FormattedMessage` using the Rich Text feature: https://formatjs.io/docs/react-intl/components#rich-text-formatting

renchap committed:
Add a new `i18n:extract` yarn command to extract localisation data

renchap committed:
Fix loading of messages in `public.jsx`

renchap commented:
@Quenty31 I see you contributed to the Occitan translation before, do you have an opinion on the custom plural issue I described above?

renchap commented:
Added polyfills for `Intl`. See the commit message for details.

renchap committed:
`babel-plugin-formatjs` needs to be in prod dependencies

renchap committed:
Properly polyfill Intl

The `intl` package has been deprecated, so let's replace it with FormatJS polyfills.

It also allows to properly load the locale data for every locale, before it always loaded `en` from the polyfill.
This created a Webpack chunk for every locale, but this is needed if we want per-locale loading.

I chose to only enable the polyfill for `Intl.PluralRules`, as the browser support for the others is good noaways. `RelativeTimeFormat` will need to use its own polyfill once we use it in the codebase, but as its not used for now, its commented.
I left the other polyfills commented out, if we notice they are really needed somehow.

The change in `tsconfig.json` is needed for dynamic imports, it has no impact for Webpack/Babel

brawaru commented:
Doesn't seem that `oc` (Occitan) and `sa` (Sanskrit) languages have plural rules defined:
- https://github.com/unicode-org/cldr-json/tree/main/cldr-json/cldr-numbers-full/main/oc
- https://github.com/unicode-org/cldr-json/tree/main/cldr-json/cldr-numbers-full/main/sa

They maybe can contribute the rules to CLDR, otherwise this warning about locale data is probably OK to ignore. It shouldn't cause any issues and Intl should resort to using default locale rules.

renchap commented:
After more testing:

- Safari supports `oc` and `sa` locales
- Firefox only supports `sa` locale
- Chrome does not support `sa` and `oc` at all
- 
I think we are fine with silencing the error in production, and relying on the browser to pick the best relevant supported locale.

renchap committed:
Only displays a warning when the browser does not advertise support for our locale

For example, `oc` locale in Google Chrome

When this happens, the browser will select another "best" locale.

renchap committed:
Remove `manage:translations` and associated files

Those no longer works for the newer `react-intl` (and last commit was 5 years ago)

renchap committed:
Update I18n check workflow

It will now extract the JS translations, and exit if it changed any files, meaning the author forgot to run `yarn i18n:extract` to handle I18n key changes

renchap commented:
Nearly done! The last commits handle the non-supported locales as explained in the initial post.

`yarn manage:translations` has been removed as it is no longer working and very outdated, we should be able to work without it using the new formatjs CLI. Documentation needs to be updated after the change.

The last remaining issue is that extraction is configured to write into `en.json`, but the default messages do not match the current english translation, so it overrides them. Maybe keys & default messages needs to be extracted to `defaultMessages.json` instead?

renchap committed:
Adjust formatjs formatted to match current behavior

renchap commented:
One more try:
- removed `defaultMessages.js` as it is not currently used anywhere (CrowdIn ingests the `en.json` file)
- matched current behaviour for I18n keys extraction: if a key exists in `en.json`, then we use the translation defined in this file, not the extracted `defaultMessage`

Running `yarn i18n:extract` with my latest commit does not change `en.json` at all, which seems correct. Adding a new key in a JS file correctly adds the key to `en.json`.

renchap committed:
Ensure CI fails if `formatjs extract` reports any issue

For example, keys with 2 different `defaultMessage`

renchap commented:
`formatjs extract` (correctly) reported that some keys were defined with different `defaultMessage` at different places in the codebase.

This is now fixed in the last commit, and I changed the GH:A workflow so it will fail if `formatjs extract` detect any issues.

This PR should be now ready for review!

renchap committed:
Replace `injectIntl` by `useIntl` in function components

ClearlyClaire commented:
It appears `app/views/layouts/embedded.html.haml` still needs to get updated.

I'd also possibly move e219447bdab51a0e1d866f4b036675d4ff244fe2 to another PR

renchap committed:
Fix locales loading in embedded layout

renchap commented:
> It appears `app/views/layouts/embedded.html.haml` still needs to get updated.

Done

> I'd also possibly move [e219447](https://github.com/mastodon/mastodon/commit/e219447bdab51a0e1d866f4b036675d4ff244fe2) to another PR

#25114

ThisIsMissEm commented on app/javascript/mastodon/polyfills/intl.ts:
```diff
@@ -0,0 +1,105 @@
+// import { shouldPolyfill as shouldPolyfillCanonicalLocales } from '@formatjs/intl-getcanonicallocales/should-polyfill';
```
@renchap does this file come from somewhere? Kinda curious about all the commented code here?

renchap replied:
No, I wrote it, but as explain in the relevant commit (I think) I wanted to let those here in case we notice they need to be enabled because we have user reporting it does not work in a given (old) browser.

I spent some time writing this and I did not wanted to redo it if it was needed :)

ThisIsMissEm reviewed (commented):
I've reviewed this code, and it looks good to me, though I might not be aware of everything that this affects. Just one or two minor comments.

renchap committed:
Display a deprecation message when `manage:translations` is ran

ClearlyClaire reviewed (approved):
It will likely be a nightmare for me to deal with in glitch-soc, but it's a welcome modernization, `injectIntl` has made things more complicated more than once.

The PR looks fine to me, but since it's a large-scale change, I'm waiting for a ack from @Gargron before merging.

ClearlyClaire commented:
Note that this breaks cross-origin loading of the assets because of the missing `crossorigin` attribute on the preload tags.
